lib/ansible/plugins/inventory/vmware_host_inventory.py

```diff
@@ -0,0 +1,514 @@
+                    cacheable_results[tag_obj.name] = {'hosts': []}
+                    self.inventory.add_group(tag_obj.name)
+
+        for tmp_host_obj in objects:
+            for tmp_host_obj_property in tmp_host_obj.propSet:
+                # VMware does not provide a way to uniquely identify host system by its name
+                # i.e. there can be two host systems with same name
+                # Appending "_" and VMware UUID to make it unique
+                current_host = tmp_host_obj_property.val + "_" + tmp_host_obj.obj.summary.hardware.uuid
+
+                if current_host not in hostvars:
+                    hostvars[current_host] = {}
+                    self.inventory.add_host(current_host)
+                    host_ip = self.get_host_ip_address(tmp_host_obj.obj)
+                    if host_ip:
+                        self.inventory.set_variable(current_host, 'ansible_host', host_ip)
+
+                    # Load Host properties in host_vars
+                    host_properties = self.get_option('properties') or []
+                    for host_prop in host_properties:
+                        host_value = self._get_host_prop(tmp_host_obj.obj, host_prop.split("."))
+                        self.inventory.set_variable(current_host, host_prop, host_value)
+
+                    # Only gather facts related to tag if vCloud and vSphere is installed.
+                    if HAS_VCLOUD and HAS_VSPHERE and self.with_tags:
+                        # Add virtual machine to appropriate tag group
+                        host_mo_id = tmp_host_obj.obj._GetMoId()
+                        host_dynamic_id = DynamicID(type='HostSystem', id=host_mo_id)
+                        attached_tags = tag_association.list_attached_tags(host_dynamic_id)
+
+                        for tag_id in attached_tags:
+                            self.inventory.add_child(tags_info[tag_id], current_host)
+                            cacheable_results[tags_info[tag_id]]['hosts'].append(current_host)
+
+                    # Based on power state of host system
+                    host_power = tmp_host_obj.obj.summary.runtime.powerState
+                    if host_power not in cacheable_results:
+                        cacheable_results[host_power] = []
+                        self.inventory.add_group(host_power)
+                    cacheable_results[host_power].append(current_host)
+                    self.inventory.add_child(host_power, current_host)
+
+        return cacheable_results
+
+    @staticmethod
+    def _get_host_prop(host, attributes):
+        """Safely get a property or return None"""
+        result = host
+        for attribute in attributes:
+            try:
+                result = getattr(result, attribute)
+            except (AttributeError, IndexError):
+                return None
+        return result
+
+    def _get_managed_objects_properties(self, vim_type, properties=None):
```

Comment:
Some of this code looks looks the same in vmware_vm_inventory, might be worth moving into common code?

Reply:
OK.